pkg/git/editor.go

```diff
@@ -0,0 +1,144 @@
+package git
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/kballard/go-shellquote"
+)
+
+/*
+NewEditor creates a new Editor instance, which can be used to launch a file (containing the content that is passed into NewEditor)
+with the OS's default IDE, and return the new file content after the user saves & closes the file.
+Accepts the name for the temporary file, and the file content.
+The name of the file will have a random string appended, or if an empty string is provided, the name will just be that random string. If the filename string includes a "*", the random string replaces the last "*".
+*/
+func NewEditor(name string, content []byte) (editor *Editor, err error) {
+	defaultIDE, err := getDefaultEditor()
+	if err != nil {
+		return nil, err
+	}
+
+	filename, err := createTemporaryFile(name, content)
+	if err != nil {
+		return nil, err
+	}
+
+	editor = &Editor{
+		Program: defaultIDE,
+		File:    filename,
+	}
+
+	return
+}
+
+/*
+Editor can be used to allow the user to directly edit content using
+their default IDE (set via `git var GIT_EDITOR`).
+*/
+type Editor struct {
+	Program string
+	File    string
+	Launch  func() error
+}
+
+/*
+EditContent launches git's default IDE and waits for the temporary file to be
+saved & closed before returning the edited file content.
+*/
+func (e *Editor) EditContent() ([]byte, error) {
+	if err := e.openAndWaitForTextEditor(); err != nil {
+		return nil, err
+	}
+
+	defer cleanup(e.File)
+
+	return readTemporaryFile(e.File)
+}
+
+func (e *Editor) openAndWaitForTextEditor() error {
+	editCmd, err := e.getOpenEditorCommand()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Waiting for your editor to close the file...")
+	if err := editCmd.Run(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Returns the command we can run to open the default git editor.
+func (e *Editor) getOpenEditorCommand() (*exec.Cmd, error) {
+	programArgs, err := shellquote.Split(e.Program)
+	if err != nil {
+		return nil, err
+	}
+
+	editCmd := exec.Command(programArgs[0], programArgs[1:]...)
+	editCmd.Stdout = os.Stdout
+	editCmd.Stdin = os.Stdin
+	editCmd.Args = append(editCmd.Args, e.File)
+	return editCmd, nil
+}
+
+// Get's the OS default git editor via the `git var GIT_EDITOR` command.
+func getDefaultEditor() (string, error) {
+	output, err := exec.Command("git", "var", "GIT_EDITOR").Output()
```

Comment:
will this work on a new docker container? do we need to check if `git` is installed first?

Reply:
good point, this will error out if git is not installed. Could fallback to [this](https://github.com/stripe/stripe-cli/pull/992#discussion_r1007101780) if it fails?

Reply:
fixed in https://github.com/stripe/stripe-cli/pull/992/commits/4932758ac2671f879224c9a3a4a35720c9aaac39